src/olympia/bandwagon/views.py

```diff
@@ -676,16 +674,49 @@ def get_queryset(self):
             author=self.get_account_viewset().get_object())
 
 
-class CollectionAddonViewSet(ListModelMixin, GenericViewSet):
-    permission_classes = [AllowAny]
+class CollectionAddonViewSet(AddonChildMixin, ModelViewSet):
+    permission_classes = []  # We don't need extra permissions.
     serializer_class = CollectionAddonSerializer
+    lookup_field = 'addon'
+    lookup_url_kwarg = 'addon_id'
 
-    def get_queryset(self):
-        if not hasattr(self, 'collection_object'):
-            self.collection_object = CollectionViewSet(
+    def get_collection_viewset(self):
+        if not hasattr(self, 'collection_viewset'):
+            # CollectionViewSet's permission_classes are good for us.
+            self.collection_viewset = CollectionViewSet(
                 request=self.request,
                 kwargs={'user_pk': self.kwargs['user_pk'],
-                        'slug': self.kwargs['collection_slug']}).get_object()
+                        'slug': self.kwargs['collection_slug']})
+        return self.collection_viewset
+
+    def get_addon_object(self):
+        """Return addon object associated with the request, or None if not
+        relevant.
+
+        Will also fire permission checks on the addon object when it's loaded.
+        """
+        if not hasattr(self, 'addon_object'):
+            if self.lookup_url_kwarg not in self.kwargs:
+                addon_id = self.request.data.get(self.lookup_url_kwarg)
+                if addon_id:
+                    self.kwargs[self.lookup_url_kwarg] = unicode(addon_id)
+            if self.lookup_url_kwarg in self.kwargs:
+                # When loading the add-on, pass a specific permission class -
+                # the default from AddonViewSet is too restrictive.
+                self.addon_object = (
+                    super(CollectionAddonViewSet, self).get_addon_object(
+                        permission_classes=[AllowIfPublic],
+                        lookup=self.lookup_url_kwarg))
+            else:
+                self.addon_object = None
+        return self.addon_object
 
+    def get_queryset(self):
         return CollectionAddon.objects.filter(
-            collection=self.collection_object)
+            collection=self.get_collection_viewset().get_object())
+
+    def create(self, request, *args, **kwargs):
+        # Drop 'addon' from data if provided - it makes the serializer choke.
+        request.data.pop('addon', None)
```

Comment:
I think it should be a 400 if provided in data (though with a better error than the one you're getting now), instead of silently ignored - to make it obvious what's going on. It should also be tested.

But see my comment above in the documentation, I think we should accept `addon` instead of `addon_id`. That would remove the need for this hack.